### PR TITLE
Clean ghost containers on startup

### PR DESCRIPTION
**Pull request by kursatkobya**
Add function to clean ghost containers on startup.
Add service tests for this feature .

### CODE REVIEW

File: agent/src/main.cpp [hunk 1]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
     return 0;
 }
 
+void cleanGhostContainers(void)
```
Comment (breakreturn) [inline agent/src/main.cpp]: I don't like this name that much. Could it be more descriptive? E.g. removeOldContainers or something like that.

File: agent/src/main.cpp [hunk 2]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
...
+{
+    char **containers = nullptr;
+    struct lxc_container **cret = nullptr;
```
Comment (breakreturn) [inline agent/src/main.cpp]: "cret"? A more descriptive name would be nice.

File: agent/src/main.cpp [hunk 3]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
...
+    const char * basePath = lxc_get_global_config_item("lxc.lxcpath");
```
Comment (breakreturn) [inline agent/src/main.cpp]: The * goes next to the variable name. Goes for a few other places in the file as well.

File: agent/src/main.cpp [hunk 4]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
...
+    auto num = list_all_containers(basePath, &containers, &cret);
+    if (num > 0 && num < std::numeric_limits<int>::max()) {
+        log_warning() << num << " ghost containers found";
```
Comment (breakreturn) [inline agent/src/main.cpp]: I think we should describe what happens better, "ghost" is perhaps better explained as "previously started", or some other wording that explains what is going on.

File: agent/src/main.cpp (hunk 1 above)
Comment (breakreturn) [inline agent/src/main.cpp]: Comments for this method would be nice.

File: agent/src/main.cpp [hunk 5]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
...
+            delete containers[num-1];
+        }
+        delete cret;
+        delete containers;
+    }
+}
 
 int main(int argc, char **argv)
 {
+    try {
```
Comment (breakreturn) [inline agent/src/main.cpp]: What's the reason we do this in main and not in the agent?

File: servicetest/agent/test_agent.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+# Test suites #
+
+
+@pytest.mark.usefixtures("agent")
+class TestClearGhostContainers(object):
+    """ This test suite is stands for testing agent to examine cleaning
+        ghost containers feature
+    """
+
+    def testme(self):
```
Comment (breakreturn) [inline servicetest/agent/test_agent.py]: This is not a good test name.

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: agent/src/main.cpp (hunk 1 above)
Comment (kursatkobya) [reply agent/src/main.cpp]: sure

File: agent/src/main.cpp [hunk 7]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
...
+void cleanGhostContainers(void)
+{
+    char **containers = nullptr;
+    struct lxc_container **cret = nullptr;
+    const char * basePath = lxc_get_global_config_item("lxc.lxcpath");
+    auto num = list_all_containers(basePath, &containers, &cret);
+    if (num > 0 && num < std::numeric_limits<int>::max()) {
+        log_warning() << num << " ghost containers found";
+        for (; num; num--) {
+            log_debug() << "Ghost container named " << containers[num-1] << " will be deleted";
```
Comment (tobsan) [inline agent/src/main.cpp]: Instead of indexing in the array on every access, couldn't you save that as a variable inside the loop instead?

`struct lxc_container *container = containers[num-1];`

File: agent/src/main.cpp [hunk 8]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
...
+
+            const char * state = cret[num-1]->state(cret[num-1]);
+            if ("RUNNING" == std::string(state)) {
```
Comment (tobsan) [inline agent/src/main.cpp]: Use `is_running()` from `lxccontainer.h` instead if you only want to check for running and don't care about other states.

File: agent/src/main.cpp [hunk 9]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
...
+                bool success = cret[num-1]->stop(cret[num-1]);
```
Comment (tobsan) [inline agent/src/main.cpp]: Dunno if it would make more sense to use shutdown() before calling stop(). What do you think?

File: agent/src/main.cpp [hunk 10]
```diff
@@ -52,9 +53,51 @@ int signalHandler(void *data) {
...
+
+            bool success = cret[num-1]->destroy(cret[num-1]);
+            if (!success) {
+                std::string errorMsg = "Unable to destroy ghost container " +
+                                       std::string(containers[num-1]);
+                throw SoftwareContainerAgentError(errorMsg);
+            }
+
+            log_debug() << "Ghost container " << containers[num-1] << " is successfully destroyed";
+            delete cret[num-1];
```
Comment (tobsan) [inline agent/src/main.cpp]: I think you should use `lxc_container_put()` instead

File: agent/src/main.cpp (hunk 5 above)
Comment (tobsan) [reply agent/src/main.cpp]: Yes, this should be done in the agent. Partly because one might have configured to put containers in a non-standard location, which one would need to know when getting all containers from lxc.

**Review by tobsan — CHANGES_REQUESTED** [review]

File: agent/src/main.cpp (hunk 7 above)
Comment (kursatkobya) [reply agent/src/main.cpp]: i could do but, this way seem better to me. Do you object for readability concerns ?

Comment (tobsan) [reply agent/src/main.cpp]: Yes, this is hard to read.

File: agent/src/main.cpp (hunk 9 above)
Comment (kursatkobya) [reply agent/src/main.cpp]: Running containers cannot be destroyed before stopping. But after stopping they can be destroyed. I also tested this scenario and it works this way.

File: agent/src/main.cpp (hunk 10 above)
Comment (kursatkobya) [reply agent/src/main.cpp]: instead of delete ? is that function frees the resource ?

File: agent/src/main.cpp (hunk 5 above)
Comment (kursatkobya) [reply agent/src/main.cpp]: I thought this operation is earlier initialization process before creating agent object. That's why it is implemented there. Surely it can be moved into agent.

File: agent/src/main.cpp (hunk 9 above)
Comment (tobsan) [reply agent/src/main.cpp]: Yes, but there are two ways to stop a container before destroying it: shutdown() and stop(), both are on the LXC API. shutdown() is the nicer version, and stop() is the SIGKILL version. Please use shutdown() as the first measure, and then use stop() if shutdown() returns false.

File: agent/src/main.cpp (hunk 10 above)
Comment (tobsan) [reply agent/src/main.cpp]: Just noticed that the API explicitly says that you should free the args, so never mind my first comment :)

File: agent/src/softwarecontaineragent.cpp [hunk 11]
```diff
@@ -91,6 +94,47 @@ SoftwareContainerAgent::~SoftwareContainerAgent()
...
+void SoftwareContainerAgent::removeOldContainers()
+{
+    char **containerNames = nullptr;
+    struct lxc_container **containerList = nullptr;
+
+    const char *basePath = lxc_get_global_config_item("lxc.lxcpath");
+    auto num = list_all_containers(basePath, &containerNames, &containerList);
+    if (num > 0 && num < std::numeric_limits<int>::max()) {
+        log_warning() << num << " unused deprecated containers found";
+        for (; num; num--) {
```
Comment (tobsan) [inline agent/src/softwarecontaineragent.cpp]: This type of truth checking is not nice. I would prefer a regular for loop with a new variable which you can compare to num. That way, you also can get rid of the `if (num > 0)` check above I guess.

TL;DR: Numbers are not bools :)

File: servicetest/agent/test_agent.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+            The second phase of the test is running agent again to see
+            ghost container is cleaned.
+        """
+        sc = Container()
+        container_id = sc.start(DATA)
+        os.system("pkill --signal SIGKILL -f softwarecontainer-agent")
+        time.sleep(0.5)
+
+        os.system("lxc-ls >> output")
```
Comment (tobsan) [inline servicetest/agent/test_agent.py:83]: It would be nice to be able to use the LXC Python binding here. Dunno if that would mean lots of more work, but it does feel more robust than parsing the lxc-ls output.

**Review by tobsan — CHANGES_REQUESTED** [review]

File: servicetest/agent/test_agent.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+from testframework import Container, SoftwareContainerAgentHandler
+
+
+CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
+
+
+# Provide what the testframework requires #####
+
+# This function is used by the 'agent' fixture to know where the log should be stored
+def logfile_path():
```
Comment (breakreturn) [inline servicetest/agent/test_agent.py:34]: We don't seem to ignore this logfile for git anywhere. I propose we follow the same "testoutput dir" pattern as most of the other tests do, and put the log file there, and ignore the directory in .gitignore

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: agent/src/softwarecontaineragent.cpp (hunk 11 above)
Comment (kursatkobya) [reply agent/src/softwarecontaineragent.cpp]: checking upper and lower limits wouldn't hurt, contrarily it helps a lot in aspect of secure code imo. Yet i can add another index variable if it is really needed.

Comment (kursatkobya) [reply agent/src/softwarecontaineragent.cpp]: ah you were talking about   `for (; num; num--)` 
aye aye sir

File: servicetest/agent/test_agent.py (hunk 12 above)
Comment (kursatkobya) [reply servicetest/agent/test_agent.py:83]: let me try this

**Review by breakreturn — APPROVED** [review]

File: agent/src/softwarecontaineragent.cpp [hunk 14]
```diff
@@ -18,6 +18,7 @@
  */
 
 #include <cstdint>
+#include <lxc/lxccontainer.h>
```
Comment (tobsan) [inline agent/src/softwarecontaineragent.cpp:21]: Could you add so that we link against LXC in the CMakeLists.txt for the agent as well?

File: agent/src/softwarecontaineragent.cpp [hunk 15]
```diff
@@ -91,6 +94,47 @@ SoftwareContainerAgent::~SoftwareContainerAgent()
...
+/* we are fetching lxcpath from config now. But in near future we should use lxcConfigPath instead
+ * */
+void SoftwareContainerAgent::removeOldContainers()
+{
+    char **containerNames = nullptr;
+    struct lxc_container **containerList = nullptr;
+
+    const char *basePath = lxc_get_global_config_item("lxc.lxcpath");
+    auto num = list_all_containers(basePath, &containerNames, &containerList);
+    if (num > 0 && num < std::numeric_limits<int>::max()) {
```
Comment (tobsan) [inline agent/src/softwarecontaineragent.cpp]: Wouldn't you rather want to change this to check for -1 (the error case) and otherwise just use the for loop to iterate over the found containers?

**Review by tobsan — CHANGES_REQUESTED** [review]

File: agent/src/softwarecontaineragent.cpp (hunk 15 above)
Comment (kursatkobya) [reply agent/src/softwarecontaineragent.cpp]: if we are trusting the library that we are sure that `list_all_containers` only return -1  on error and count of containers on success, then you are right. This check is questioning borders instead of error check, which is safer. I didn't look in `list_all_containers` code so i don't know whether it has checks or not.

File: agent/src/softwarecontaineragent.cpp [hunk 16]
```diff
@@ -91,6 +94,50 @@ SoftwareContainerAgent::~SoftwareContainerAgent()
...
+void SoftwareContainerAgent::removeOldContainers()
+{
+    char **containerNames = nullptr;
+    struct lxc_container **containerList = nullptr;
+
+    const char *basePath = lxc_get_global_config_item("lxc.lxcpath");
+    auto num = list_all_containers(basePath, &containerNames, &containerList);
+
+    if (-1 == num) {
+        throw SoftwareContainerAgentError("An error is occurred while trying to get container list");
```
Comment (breakreturn) [inline agent/src/softwarecontaineragent.cpp:114]: Probably good to log here as well.

File: agent/src/softwarecontaineragent.cpp [hunk 17]
```diff
@@ -91,6 +94,51 @@ SoftwareContainerAgent::~SoftwareContainerAgent()
...
+void SoftwareContainerAgent::removeOldContainers()
+{
+    char **containerNames = nullptr;
+    struct lxc_container **containerList = nullptr;
+
+    const char *basePath = lxc_get_global_config_item("lxc.lxcpath");
+    auto num = list_all_containers(basePath, &containerNames, &containerList);
+
+    if (-1 == num) {
+        log_warning() << "An error is occurred while trying to get deprecated container list";
```
Comment (tobsan) [inline agent/src/softwarecontaineragent.cpp]: This should be log_error().

**Review by tobsan — APPROVED** [review]
Looks good, just had one minor comment
